### DISCUSSION TITLE
Wrong Certificate-Bound Refresh Tokens

### DISCUSSION THREAD

**Discussion by pedroh-pereira** [keycloak/keycloak#19704] [Q&A] [answered]
The certificate-bound part was implemented based on a [draft](https://tools.ietf.org/html/draft-ietf-oauth-token-binding-08) that became [RFC 8705 OAuth 2.0 Mutual-TLS Client Authentication and Certificate-Bound Access Tokens
](https://www.rfc-editor.org/rfc/rfc8705)
We are testing FAPI 1 advanced profile and facing problems with certificate-bound to refresh tokens. According to [RFC 8705 - 7.1 Certificate-Bound Refresh Tokens](https://www.rfc-editor.org/rfc/rfc8705#name-certificate-bound-refresh-t) a refresh token is already linked to client indirectly when the DCR param `token_endpoint_auth_method` is `tls_client_auth` (but on FAPI 1 advanced profile all requests must be MTLS, so Keycloak validates all requests based on the certificate no matters the `token_endpoint_auth_method`). So that, the refresh token must not have `cnf` claim for confidential clients, because if a client updates the certificate it'll invalidate the refresh token, since keycloak validates this claim and according to [RFC 8705 - 6.3 Certificate Expiration and Bound Access Tokens](https://www.rfc-editor.org/rfc/rfc8705.html#name-certificate-expiration-and-)  when this happens the access token bounded to old certificate should be treated similarly to expired token, as a result of that the client should generate a new access using the refresh token that is not possible in the latest keycloak version.

This feature was implemented by the following issues: [KEYCLOAK-6767](https://issues.redhat.com/browse/KEYCLOAK-6767) and [KEYCLOAK-6771](https://issues.redhat.com/browse/KEYCLOAK-6771)

**How to reproduce the error**

Certificates:
Both certificates are using the same attributes, but different expiration dates.
-  CertificateA - Expiration time 2023/12/01
- CertificateB - Expiration time 2024/12/01

Steps:
1 - Register a client with `tls_client_auth` using CertificateA
2 - Generate a access token using `authorization_code` grant type
3 - Store this refresh_token
4 - Execute refresh token flow using Certificate A (It'll work)
5 - Execute refresh token flow using Certificate B (It won't work and it'll return a 401 Unauthorized)

Places containing refresh token validations:
![image](https://user-images.githubusercontent.com/47221187/231782001-68884ac6-434a-47c8-af14-d8e2fd00da59.png)
![image](https://user-images.githubusercontent.com/47221187/231786114-22226f52-4828-4906-9552-15a48ae7f193.png)
![image](https://user-images.githubusercontent.com/47221187/231786777-54218851-706a-46e8-b0c3-500ab18924b6.png)

**Comment by jlosantana** [comment]
Very good explanation @pedroh-pereira! I believe it is a lack of rfc tracking problem, since we saw a comment like this:

`// KEYCLOAK-6771 Certificate Bound Token
 // https://tools.ietf.org/html/draft-ietf-oauth-mtls-08#section-3.1`

https://github.com/keycloak/keycloak/pull/5083/commits/54a0ce8e60afa7d171f4cb68000df6904dfd29dd

This commit made on Jun 5, 2018, and RFC8705 was published on Feb, 2020. Maybe we need to update this implementation.

**Comment by tnorimat** [comment, marked as answer]
@pedroh-pereira Hello, thank you for noticing the issue. I will consider this issue. Yes, I implemented the certificate-bound token by following the internet draft version of OAuth MTLS.

**Reply by pedroh-pereira** [reply]
No problem, thanks for your attention!
